// @angular/platform-server/platform_browser_private.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
"use strict";
var platform_browser_1 = require('@angular/platform-browser');
exports.DomAdapter = platform_browser_1.__platform_browser_private__.DomAdapter;
exports.setRootDomAdapter = platform_browser_1.__platform_browser_private__.setRootDomAdapter;
exports.getDOM = platform_browser_1.__platform_browser_private__.getDOM;
exports.DomRootRenderer = platform_browser_1.__platform_browser_private__.DomRootRenderer;
exports.DomRootRenderer_ = platform_browser_1.__platform_browser_private__.DomRootRenderer_;
exports.DomSharedStylesHost = platform_browser_1.__platform_browser_private__.DomSharedStylesHost;
exports.SharedStylesHost = platform_browser_1.__platform_browser_private__.SharedStylesHost;
exports.ELEMENT_PROBE_PROVIDERS = platform_browser_1.__platform_browser_private__.ELEMENT_PROBE_PROVIDERS;
exports.DomEventsPlugin = platform_browser_1.__platform_browser_private__.DomEventsPlugin;
